### PR TITLE
sdio: add sdio register definition, functions and testcases.

### PR DESCRIPTION
**Pull request by NanaHigh**
- there are only a few implementation of ambiguous registers has not yet been implemented.

### CODE REVIEW

File: bouffalo-hal/src/sdio.rs [hunk 1]
```diff
@@ -66,151 +65,3473 @@ pub struct RegisterBlock {
...
+/// Data Transfer Mode.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub enum DataTransferMode {
+    /// Other modes.
+    Other,
+    /// Master in, slave out.
+    MISO,
+}
+
+/// Auto CMD Mode
```
Comment (luojia65) [inline bouffalo-hal/src/sdio.rs:230]:
```suggestion
/// Auto CMD Mode.
```

File: bouffalo-hal/src/sdio.rs [hunk 2]
```diff
@@ -66,151 +65,3473 @@ pub struct RegisterBlock {
...
+    #[inline]
+    pub fn enable_dma(self) -> Self {
+        Self((self.0 & !Self::DMA_ENABLE) | Self::DMA_ENABLE & 1)
+    }
+    /// Disable DMA.
+    #[inline]
+    pub fn disable_dma(self) -> Self {
+        Self((self.0 & !Self::DMA_ENABLE) | Self::DMA_ENABLE & 0)
+    }
+    /// If DMA is enabled.
```
Comment (luojia65) [inline bouffalo-hal/src/sdio.rs:310]:
```suggestion
    /// Checks if DMA is enabled.
```

File: bouffalo-hal/src/sdio.rs [hunk 3]
```diff
@@ -66,151 +65,3473 @@ pub struct RegisterBlock {
...
+            3 => ResponseType::ResponseLen48Check,
+            _ => ResponseType::NoResponse,
+        }
+    }
+}
+
+/// Register that stores responses from SD cards.
 #[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
 #[repr(transparent)]
 pub struct Response(u128);
```
Comment (luojia65) [inline bouffalo-hal/src/sdio.rs:444]: We should check if reading and writing of `u128` values on Rust are supported in the target peripheral later.

File: bouffalo-hal/src/sdio.rs [hunk 4]
```diff
@@ -66,151 +65,3473 @@ pub struct RegisterBlock {
...
 
+impl Response {
+    /// Set Command Response.
```
Comment (luojia65) [inline bouffalo-hal/src/sdio.rs:447]:
```suggestion
    /// Set command response.
```

**Review by luojia65 — APPROVED** [review]
Nice! The code generally LGTM, we can apply some code style fixes, though.

**Comment by NanaHigh** [comment]
@luojia65 I took the first two suggestions and modified the commit. But about the last case-based suggestions, it's because I directly copied many names on the datasheet as comments. It will take a lot of time to check the changes one by one, can I leave them as they are?

**Comment by luojia65** [comment]
@NanaHigh : Thank you! Yes you can just leave the documentation as-is. We'll head on to modify it later if necessary.

**Pull request merged**